WreckEngine/Text.h

```diff
@@ -9,63 +9,148 @@
-		FT_Face fontFace = nullptr;
-		uint32_t height = 0, width = 0;
-		std::unordered_map<char, glyph> glyphs;
-		static FT_Long loadedFonts;
-
-		void loadGlyphs();
-		void setSize(const uint32_t height, const uint32_t width = 0);
-	};
-
-	struct Instance {
-		std::string text;
-		const FontFace* font;
-		float x, y, scale;
-		vec4 color;
-	};
-
-	struct Renderer {
-		void init();
-		// x and y are in pixels
-		void draw(Instance& inst);
+    using glm::uvec4;
+
+    void init();
+    extern bool active;
+
+    struct FontFace;
+    shared<FontFace> loadFont(const std::string& font, const uint32_t height, const uint32_t width = 0);
+    shared<FontFace> loadWinFont(const std::string& font, const uint32_t height, const uint32_t width = 0);
+
+    struct Glyph {
+        vec2 bearing;
+        vec2 size;
+        vec2 texBearing;
+        vec2 texSize;
+        float advance;
+    };
+
+    struct FontFace {
+        static constexpr int TEX_SIZE = 512;
+
+        FontFace() = default;
+        FontFace(const std::string& font); ~FontFace();
+
+        FT_Face fontFace = nullptr;
+        uint32_t height = 0, width = 0;
+        float spaceWidth = 0.0f;
+        float lineHeight = 0.0f;
+        GLtexture tex;
+        std::unordered_map<uint32_t, Glyph> glyphs;
+
+        void loadGlyphs();
+        void loadGlyphRange(uint32_t begin, uint32_t end);
+        void setSize(const uint32_t height, const uint32_t width = 0);
+    };
+
+    enum Justify : GLubyte {START, MIDDLE, END};
+
+    vec2 getDims(char ch, const FontFace* font, float scale);
+    vec2 getDims(uint32_t cp, const FontFace* font, float scale);
+    vec2 getDims(const std::string& text, const FontFace* font, float scale);
+    vec2 getDims(const std::string& text, const FontFace* font, float scale, int& lineCount);
+    float getKerning(char ch1, char ch2, const FontFace* font);
+    float getKerning(uint32_t cp1, uint32_t cp2, const FontFace* font);
+
+    struct Instance {
+        Instance();
+        
+        void queueForDraw();
+
+        inline void alignHorizontal(Justify justify) {
+            if (horiz != justify) {
+                horiz = justify;
+                dirtyAlign = true;
+            }
+        }
+        inline void alignVertical(Justify justify) {
+            if (vert != justify) {
+                vert = justify;
+                dirtyAlign = true;
+            }
+        }
+        inline void setColor(const vec4& _color) {
+            color = _color;
+            dirtyBuffer = true;
+        }
+        inline void setFont(FontFace* _font) {
+            if (font != _font) {
+                font = _font;
+                if (font) {
+                    renderInfo.setTextures(font->tex);
+                }
+                dirtyBuffer = true;
+            }
+        }
+        inline void setOffset(const vec2& _offset) {
+            offset = _offset;
+            fullOffset.value = offset + alignOffset;
```

Comment:
I think this is a line I added when making this addition that became irrelevant after I moved it into the update loop. It's unnecessary now, you should delete it